extensions/amp-autocomplete/0.1/amp-autocomplete.js

```diff
@@ -291,33 +315,48 @@ export class AmpAutocomplete extends AMP.BaseElement {
   }
 
   /**
-   * Render filtered results on the current input and update the container_.
+   * Filter the source data according to the given input and render it in
+   * the results container_.
+   * @param {Array<!JsonObject|string>=} sourceData
+   * @param {string=} input
    * @return {!Promise}
    * @private
    */
-  renderResults_() {
+  filterDataAndRenderResults_(sourceData, input = '') {
     this.clearAllItems_();
-    if (this.userInput_.length < this.minChars_ || !this.inlineData_) {
+    if (input.length < this.minChars_ || !sourceData || !sourceData.length) {
       return Promise.resolve();
     }
+    const filteredData = this.filterData_(sourceData, input);
+    return this.renderResults_(filteredData, this.container_);
+  }
 
-    const filteredData = this.filterData_(this.inlineData_, this.userInput_);
+  /**
+   * Render the given data into item elements in the given container element.
+   * @param {!Array<!JsonObject|string>} filteredData
+   * @param {?Element} container
```

Comment:
Since render doesn't happen until after layoutCallback, and `this.container_` is always assigned a non-nullable `{!Element}` in `buildCallback`, we should specify the parameter here as `{!Element}` and at the callsite use `dev().assertElement(this.container_)` so that we don't need extraneous null checks in our methods to make Closure Compiler happy.